File: vipclaw-admin/src/main/java/com/vipamp/vipclaw/admin/controller/SysUserController.java
package com.vipamp.vipclaw.admin.controller;

import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.vipamp.vipclaw.admin.dto.SysUserCreateRequest;
import com.vipamp.vipclaw.admin.dto.SysUserResponse;
import com.vipamp.vipclaw.admin.dto.SysUserUpdateRequest;
import com.vipamp.vipclaw.admin.entity.SysUser;
import com.vipamp.vipclaw.admin.service.SysUserService;
import com.vipamp.vipclaw.admin.vo.Result;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

/**
 * 用户管理控制器
 *
 * @author vipamp
 * @since 2026-03-06
 */
@Slf4j
@RestController
@RequestMapping("/users")
@RequiredArgsConstructor
@Tag(name = "用户管理", description = "用户相关接口")
public class SysUserController {

    private final SysUserService sysUserService;

    @GetMapping("/list")
    @Operation(summary = "分页获取用户列表", description = "分页查询用户信息")
    public Result<Page<SysUserResponse>> getUserPage(
            @Parameter(description = "页码", example = "1") @RequestParam(name = "pageNum", defaultValue = "1") Integer pageNum,
            @Parameter(description = "每页大小", example = "10") @RequestParam(name = "pageSize", defaultValue = "10") Integer pageSize,
            @Parameter(description = "模糊查询字段") @RequestParam(name = "keyword", required = false) String keyword,
            @Parameter(description = "状态筛选字段") @RequestParam(name = "status", required = false) Integer status) {
        try {
            Page<SysUser> page = sysUserService.getUserPage(keyword, status, pageNum, pageSize);
            Page<SysUserResponse> responsePage = convertToResponsePage(page);
            return Result.success(responsePage);
        } catch (Exception e) {
            log.error("获取用户列表失败", e);
            return Result.error(e.getMessage());
        }
    }

    @GetMapping("/{id}")
    @Operation(summary = "获取用户详情", description = "根据用户 ID 获取用户信息")
    public Result<SysUserResponse> getUserById(
            @Parameter(description = "用户 ID") @PathVariable Long id) {
        try {
            SysUser user = sysUserService.getUserById(id);
            return Result.success(SysUserResponse.fromEntity(user));
        } catch (Exception e) {
            log.error("获取用户详情失败", e);
            return Result.error(e.getMessage());
        }
    }

    @PostMapping
    @Operation(summary = "创建用户", description = "新增用户信息")
    public Result<Void> createUser(
            @Valid @RequestBody SysUserCreateRequest request) {
        try {
            return sysUserService.createUser(request) ? Result.success() : Result.error("创建用户失败");
        } catch (Exception e) {
            log.error("创建用户失败", e);
            return Result.error(e.getMessage());
        }
    }

    @PutMapping("/update/{userId}")
    @Operation(summary = "更新用户", description = "根据用户 ID 更新用户信息")
    public Result<Void> updateUser(
            @Parameter(description = "用户 ID") @PathVariable Long userId,
            @Valid @RequestBody SysUserUpdateRequest request) {
        try {
            request.setId(userId);
            return sysUserService.updateUser(userId, request) ? Result.success() : Result.error("更新用户失败");
        } catch (Exception e) {
            log.error("更新用户失败", e);
            return Result.error(e.getMessage());
        }
    }

    @PutMapping("/toggle/{userId}")
    @Operation(summary = "更新用户", description = "根据用户 ID 更新用户信息")
    public Result<Void> toggleUser(
            @Parameter(description = "用户 ID") @PathVariable Long userId,
            @Parameter(description = "用户状态") @RequestParam(name = "status") Integer status) {
        try {
            return sysUserService.toggleUserStatus(userId, status) ? Result.success() : Result.error("更新用户失败");
        } catch (Exception e) {
            log.error("更新用户失败", e);
            return Result.error(e.getMessage());
        }
    }

    @DeleteMapping("/{userId}")
    @Operation(summary = "删除用户", description = "根据用户 ID 删除用户")
    public Result<Void> deleteUser(
            @Parameter(description = "用户 ID") @PathVariable Long userId) {
        try {
            return sysUserService.deleteUser(userId) ? Result.success() : Result.error("删除用户失败");
        } catch (Exception e) {
            log.error("删除用户失败", e);
            return Result.error(e.getMessage());
        }
    }

    /**
     * 分页结果转换
     */
    private Page<SysUserResponse> convertToResponsePage(Page<SysUser> page) {
        Page<SysUserResponse> responsePage = new Page<>(page.getCurrent(), page.getSize());
        responsePage.setTotal(page.getTotal());
        responsePage.setSize(page.getSize());
        responsePage.setCurrent(page.getCurrent());
        responsePage.setPages(page.getPages());
        responsePage.setRecords(page.getRecords().stream()
                .map(SysUserResponse::fromEntity)
                .toList());
        return responsePage;
    }
}

File: vipclaw-admin/src/main/kotlin/com/vipamp/vipclaw/admin/controller/SysUserController.kt
package com.vipamp.vipclaw.admin.controller

import com.vipamp.vipclaw.admin.config.EditionUtil
import com.vipamp.vipclaw.admin.config.RequiresEdition
import com.vipamp.vipclaw.admin.dto.Page
import com.vipamp.vipclaw.admin.dto.ResultVo
import com.vipamp.vipclaw.admin.dto.SysUserCreateRequest
import com.vipamp.vipclaw.admin.dto.SysUserResponse
import com.vipamp.vipclaw.admin.dto.SysUserUpdateRequest
import com.vipamp.vipclaw.admin.dto.mapRecords
import com.vipamp.vipclaw.admin.service.SysUserService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.*

/**
 * User management controller
 * Available only for enterprise and public editions
 */
@RestController
@RequestMapping("/api/users")
@Tag(name = "User Management", description = "User related APIs")
@RequiresEdition("enterprise", "public")
class SysUserController(
    private val sysUserService: SysUserService,
    private val editionUtil: EditionUtil,
) {

    private val log = LoggerFactory.getLogger(SysUserController::class.java)

    @GetMapping("/page")
    @Operation(summary = "Get user list with pagination", description = "Paginated query for user information")
    fun pageSysUser(
        @Parameter(description = "Page number", example = "1") @RequestParam(
            name = "pageNum",
            defaultValue = "1",
        ) pageNum: Int?,
        @Parameter(description = "Page size", example = "10") @RequestParam(
            name = "pageSize",
            defaultValue = "10",
        ) pageSize: Int?,
        @Parameter(description = "Keyword for fuzzy search") @RequestParam(name = "keyword", required = false) keyword: String?,
        @Parameter(description = "Status filter") @RequestParam(name = "status", required = false) status: Int?,
        @Parameter(description = "Tenant ID filter") @RequestParam(name = "tenantId", required = false) tenantId: Long?,
    ): ResultVo<Page<SysUserResponse>> = try {
        val page = sysUserService.page(
            keyword,
            status,
            tenantId,
            pageNum ?: 1,
            pageSize ?: 10,
        )
        ResultVo.success(page.mapRecords { sysUserService.convertToResponse(it) })
    } catch (e: Exception) {
        log.error("Failed to get user list", e)
        ResultVo.error(e.message ?: "Failed to get user list")
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get user details", description = "Get user information by user ID")
    fun getSysUser(
        @Parameter(description = "User ID") @PathVariable(name = "id") id: Long,
    ): ResultVo<SysUserResponse?> = try {
        val user = sysUserService.getSysUser(id)
        ResultVo.success(user?.let { sysUserService.convertToResponse(it) })
    } catch (e: Exception) {
        log.error("Failed to get user details", e)
        ResultVo.error(e.message ?: "Failed to get user details")
    }

    @PostMapping
    @Operation(summary = "Create user", description = "Add new user information")
    fun createUser(
        @Valid @RequestBody request: SysUserCreateRequest,
    ): ResultVo<Void> = try {
        // Pass edition information, personal edition does not require email and phone
        if (sysUserService.createUser(request, editionUtil.isPersonal())) ResultVo.success() else ResultVo.error("Failed to create user")
    } catch (e: Exception) {
        log.error("Failed to create user", e)
        ResultVo.error(e.message ?: "Failed to create user")
    }

    @PutMapping("/update/{id}")
    @Operation(summary = "Update user", description = "Update user information by user ID")
    fun updateSysUser(
        @Parameter(description = "User ID") @PathVariable(name = "id") id: Long,
        @Valid @RequestBody request: SysUserUpdateRequest,
    ): ResultVo<Void> = try {
        // Pass edition information, personal edition does not perform required field validation
        if (sysUserService.updateUser(id, request, editionUtil.isPersonal())) ResultVo.success() else ResultVo.error("Failed to update user")
    } catch (e: Exception) {
        log.error("Failed to update user", e)
        ResultVo.error(e.message ?: "Failed to update user")
    }

    @PutMapping("/toggle/{id}")
    @Operation(summary = "Toggle user status", description = "Toggle user status by user ID")
    fun toggleSysUser(
        @Parameter(description = "User ID") @PathVariable(name = "id") id: Long,
        @Parameter(description = "User status") @RequestParam(name = "status") status: Int,
    ): ResultVo<Void> = try {
        if (sysUserService.toggleUserStatus(id, status)) ResultVo.success() else ResultVo.error("Failed to toggle user status")
    } catch (e: Exception) {
        log.error("Failed to update user", e)
        ResultVo.error(e.message ?: "Failed to update user")
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Delete user", description = "Delete user by user ID")
    fun deleteSysUser(
        @Parameter(description = "User ID") @PathVariable(name = "id") id: Long,
    ): ResultVo<Void> = try {
        if (sysUserService.deleteUser(id)) ResultVo.success() else ResultVo.error("Failed to delete user")
    } catch (e: Exception) {
        log.error("Failed to delete user", e)
        ResultVo.error(e.message ?: "Failed to delete user")
    }

    @GetMapping("/check/username")
    @Operation(summary = "Check username", description = "Check if username is already registered")
    fun checkUsername(
        @Parameter(description = "Username") @RequestParam(name = "username") username: String,
    ): ResultVo<Boolean> = try {
        val exists = sysUserService.existsByUsername(username)
        ResultVo.success(exists)
    } catch (e: Exception) {
        log.error("Failed to check username", e)
        ResultVo.error(e.message ?: "Failed to check username")
    }

    @GetMapping("/check/phone")
    @Operation(summary = "Check phone", description = "Check if phone is already registered")
    fun checkPhone(
        @Parameter(description = "Phone number") @RequestParam(name = "phone") phone: String,
    ): ResultVo<Boolean> = try {
        val exists = sysUserService.existsByPhone(phone)
        ResultVo.success(exists)
    } catch (e: Exception) {
        log.error("Failed to check phone", e)
        ResultVo.error(e.message ?: "Failed to check phone")
    }

    @GetMapping("/check/email")
    @Operation(summary = "Check email", description = "Check if email is already registered")
    fun checkEmail(
        @Parameter(description = "Email") @RequestParam(name = "email") email: String,
    ): ResultVo<Boolean> = try {
        val exists = sysUserService.existsByEmail(email)
        ResultVo.success(exists)
    } catch (e: Exception) {
        log.error("Failed to check email", e)
        ResultVo.error(e.message ?: "Failed to check email")
    }
}
